### DISCUSSION TITLE
Programtically get the ILAst code for each method

### DISCUSSION THREAD

**Discussion by unkown512** [icsharpcode/ILSpy#2958] [Help] [answered]
Is there a way in the API to get the ILAst output from ILSpy?

In the UI app, you can get the ILAst by clicking each individual method/etc one at a time. However, it would be nice to get the full ILAst for all methods in one pass.

**Comment by dgrunwald** [comment, marked as answer]
Not with ILSpy; but you can use the decompiler itself as a NuGet package ([ICSharpCode.Decompiler](https://www.nuget.org/packages/ICSharpCode.Decompiler/)).
Use the `ILReader.ReadIL()` method to get the initial ILAst, then apply the transforms as needed (you can use `CSharpDecompiler.GetILTransforms()` to get the ILAst transforms normally used by the C# decompiler).

**Reply by unkown512** [reply]
Thanks! I will try this :)

**Reply by unkown512** [reply]
Hey @dgrunwald I am having trouble getting the ILAst for each method or for the entire MainMethod created from PEFile.

The issue is getting IMethod created so that you can get the type system, resolver, and decomp settings. 

The function I was looking at is located in ILAstLanguages.cs

```public override void DecompileMethod(IMethod method, ITextOutput output, DecompilationOptions options)```


Do you have a starting point that may help resolve the IMethod creation for each method in the PEFile?

**Reply by dgrunwald** [reply]
You don't need `IMethod` to get the ILAst -- note that the function you are quoting only uses the `PEFile` and `method.MetadataToken`, none of the actual type-system functionality of `IMethod`. The ILSpy code is really unusual there because ILSpy has multiple decompiler type systems -- the `IMethod method` parameter is from the "UI type system", whereas the decompilation itself uses yet another type system. This is unnecessary complexity for what you're trying to do.

You can get the list of all method definition handles in a module with `module.Metadata.MethodDefinitions`, and that should be sufficient to get the ILAst.

Alternatively, if you want to use the type system with `IMethod` (since that might be easier to use than the raw System.Reflection.Metadata), you need to create  a `DecompilerTypeSystem` instance, then iterate over `typeSystem.MainModule.TypeDefinitions` and discover all methods from there.
